ESP8266/ESP8266.cpp

```diff
@@ -82,10 +82,31 @@ bool ESP8266::dhcp(bool enabled, int mode)
         && _parser.recv("OK");
 }
 
-bool ESP8266::connect(const char *ap, const char *passPhrase)
+int ESP8266::connect(const char *ap, const char *passPhrase)
 {
-    return _parser.send("AT+CWJAP_CUR=\"%s\",\"%s\"", ap, passPhrase)
-        && _parser.recv("OK");
+    int error_code = 0;
+    bool err = false;
+
+    _parser.send("AT+CWJAP_CUR=\"%s\",\"%s\"", ap, passPhrase);
+    //if we don't get a +CWJAP and error code as a response, connection was success
+    //Note: espressif at command document says that this should be +CWJAP_CUR:<error code>
+    //but seems that at least current version is not sending it
+    //https://www.espressif.com/sites/default/files/documentation/4a-esp8266_at_instruction_set_en.pdf
+    //Also seems that ERROR is not sent
+    err = _parser.recv("+CWJAP:%d", &error_code);
+    if(err) {
+        _parser.recv("FAIL");
+        if(error_code == 1)
+            return NSAPI_ERROR_CONNECTION_TIMEOUT;
+        else if(error_code == 2)
+            return NSAPI_ERROR_AUTH_FAILURE;
+        else if(error_code == 3)
+            return NSAPI_ERROR_NO_SSID;
+        else
+            return NSAPI_ERROR_NO_CONNECTION;
+    } else {
+        return NSAPI_ERROR_OK;
```

Comment:
You should still parse the OK, not leave it in the RX buffer.

Reply:
Hang on, is +CWJAP error only? In that case you shouldn't be doing a recv on it - it will be blocking for a while waiting for it in the success case.

It seems more likely to me that you should be doing a recv("OK"), with an OOB set up for "+CWJAP:", which notes the error code and returns false(?) to say "give up waiting for the OK".

Maybe come and discuss

Reply:
Note that multiple OOBs could cover potential alternative outputs, in case they change to match the documentation.

Reply:
So to get the different error codes parsed from "+CWJAP:" I would need to create 4 OOBs since ATCmdParser only checks that OOB-prefix matches to the received AT-response and calls for the callback-function without possibility to check for the error id, and to cover also the potential future change to match the documentation there would be 8 OOBs for this.

Reply:
In the OOB callback you are allowed to go back in to the data following the OOB prefix, either with a plain read call or a recursive recv - see packet_handler.

Reply:
Ok, that makes sense, thanks. 
One thing still, in the future if they change to match the documentation, "+CWJAP_CUR:" is also used when sent the "AT+CWJAP_CUR?" -query, so it would trigger to that response also, but can be handled in the OOB since that response starts with ssid and that is enclosed by double quotation marks where the error response only starts with an integer without quotation marks.
So all in all, 2 OOBs needed, 1 for now, 1 for possible future use.